htdocs/product/class/productbatch.class.php

```diff
@@ -539,5 +539,48 @@ public static function findAll($db, $fk_product_stock, $with_qty = 0, $fk_produc
             $error = "Error ".$db->lasterror();
             return -1;
         }
+    }
+	    /**
+     * Return all batch records for a given product and warehouse with FIFO lot in array
+     *
+     *  @param	DoliDB		$db    				database object
+     *  @param  int         $fk_product         If set to a product id, get eatby and sellby from table llx_product_lot
+     *  @param	int			$idwarehouse		id of warehouse
+     *  @return array         					<0 if KO, array of infos
+     */
+    public static function findFIFO($db, $fk_product = 0, $idw = null)
+    {
+    	global $langs;
+		$ret = array();
+
+
+		$sql = "SELECT t.rowid, t.batch, ps.reel, t.qty";
+        $sql .= " FROM ".MAIN_DB_PREFIX."product_stock as ps";
+        $sql .= " LEFT JOIN ".MAIN_DB_PREFIX."product_batch as t ON t.fk_product_stock = ps.rowid";
+        $sql .= " LEFT JOIN ".MAIN_DB_PREFIX."product_lot as pl ON pl.fk_product = ps.fk_product and pl.batch = t.batch";
+		$sql .= " WHERE ps.fk_product=".$fk_product." AND ps.fk_entrepot = ".$idw." AND t.qty > 0";
+		$sql .= " ORDER BY pl.tms";
```

Comment:
product_lot is lot info table, if you want first in stock you need to sort on tms from product_batch (lot stock table)

Reply:
Hello , sorry but i leave réunion island for LYON....
we don't take the tms of productbatch table for sorting the résult of FIFO in a warehouse because when you transfered between entrepot A to entrepot B the tms of entrepot A take the last mouvement and = to tms of entrepot B enregistrement.
Also , we are in obligation to use the tms of product lot  for sort between multi line of ewarehouse X.


I think that is a bug for the transfer function, I check in version 12.0.1
and we also have bp with the stock correction ....